sharding/node/backend.go

```diff
@@ -228,53 +228,100 @@ func (s *ShardEthereum) registerTXPool(actor string) error {
 }
 
 // Registers the actor according to CLI flags. Either notary/proposer/observer.
 func (s *ShardEthereum) registerActorService(config *params.Config, actor string, shardID int) error {
-	return s.Register(func(ctx *sharding.ServiceContext) (sharding.Service, error) {
-
-		var p2p *p2p.Server
-		ctx.RetrieveService(&p2p)
-		var smcClient *mainchain.SMCClient
-		ctx.RetrieveService(&smcClient)
-		var shardChainDB *database.ShardDB
-		ctx.RetrieveService(&shardChainDB)
-
-		if actor == "notary" {
-			return notary.NewNotary(config, smcClient, p2p, shardChainDB)
-		} else if actor == "proposer" {
-			var txPool *txpool.TXPool
-			ctx.RetrieveService(&txPool)
-			return proposer.NewProposer(config, smcClient, p2p, txPool, shardChainDB.DB(), shardID)
+	var shardp2p *p2p.Server
+	if err := s.fetchService(&shardp2p); err != nil {
+		return err
+	}
+	var client *mainchain.SMCClient
+	if err := s.fetchService(&client); err != nil {
+		return err
+	}
+
+	var shardChainDB *database.ShardDB
+	if err := s.fetchService(&shardChainDB); err != nil {
+		return err
+	}
+
+	var pool *txpool.TXPool
+	if err := s.fetchService(&pool); err != nil {
+		return err
+	}
+
+	if actor == "notary" {
+		not, err := notary.NewNotary(config, client, shardp2p, shardChainDB)
+		if err != nil {
+			return fmt.Errorf("could not register notary service: %v", err)
+		}
+		return s.registerService(not)
+	} else if actor == "proposer" {
+		prop, err := proposer.NewProposer(config, client, shardp2p, pool, shardChainDB, shardID)
+		if err != nil {
+			return fmt.Errorf("could not register proposer service: %v", err)
 		}
-		return observer.NewObserver(p2p, shardChainDB.DB(), shardID)
-	})
+		return s.registerService(prop)
+	}
+	obs, err := observer.NewObserver(shardp2p, shardChainDB, shardID)
+	if err != nil {
+		return fmt.Errorf("could not register observer service: %v", err)
+	}
+	return s.registerService(obs)
 }
 
-func (s *ShardEthereum) registerSimulatorService(config *params.Config, shardID int) error {
-	return s.Register(func(ctx *sharding.ServiceContext) (sharding.Service, error) {
-		var p2p *p2p.Server
-		ctx.RetrieveService(&p2p)
-		var smcClient *mainchain.SMCClient
-		ctx.RetrieveService(&smcClient)
-		return simulator.NewSimulator(config, smcClient.ChainReader(), smcClient.SMCCaller(), p2p, shardID, 15) // 15 second delay between simulator requests.
-	})
+func (s *ShardEthereum) registerSimulatorService(actorFlag string, config *params.Config, shardID int) error {
+	// Should not trigger simulation requests if actor is a notary, as this
+	// is supposed to "simulate" notaries sending requests via p2p.
+	if actorFlag == "notary" {
+		return nil
+	}
+
+	var shardp2p *p2p.Server
+	if err := s.fetchService(&shardp2p); err != nil {
+		return err
+	}
+	var client *mainchain.SMCClient
+	if err := s.fetchService(&client); err != nil {
+		return err
+	}
+
+	// 15 second delay between simulator requests.
+	sim, err := simulator.NewSimulator(config, client, shardp2p, shardID, 15)
```

Comment:
I'm not sure if you addressed this in #216, but can you please change this to pass `15*time.Second`? 
It's not intuitive to pass 15 as `time.Duration` and know that this will be multipled by `time.Second` later. 

If this is more than a 3-5 line change, then we can address later.